### PR TITLE
Form entry api

### PR DESCRIPTION
**Pull request by wpride**

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 11 errors:

* touchforms/backend/test_form_player_api.py, line 6 - F401 'AnswerQuestionJson' imported but unused
* touchforms/backend/test_form_player_api.py, line 6 - F401 'InstanceXmlUtil' imported but unused
* touchforms/backend/test_form_player_api.py, line 6 - F401 'PromptToJson' imported but unused
* touchforms/backend/test_form_player_api.py, line 7 - F401 'RuntimeException' imported but unused
* touchforms/backend/test_form_player_api.py, line 85 - E303 too many blank lines (2)
* touchforms/backend/test_form_player_api.py, line 136 - F841 local variable 'old_submit' is assigned to but never used
* touchforms/backend/test_form_player_api.py, line 174 - E303 too many blank lines (2)
* touchforms/backend/test_form_player_api.py, line 189 - E713 test for membership should be 'not in'
* touchforms/backend/test_form_player_api.py, line 204 - E302 expected 2 blank lines, found 1
* touchforms/backend/test_form_player_api.py, line 205 - E703 statement ends with a semicolon
* touchforms/backend/xformplayer.py, line 192 - E225 missing whitespace around operator

File: touchforms/backend/test_files/xforms/repeat.xml [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+		<group>
+			<label ref="jr:itext('num_repeat-label')"/>
+			<repeat jr:count="/data/num_repeats" jr:noAddRemove="true()" nodeset="/data/num_repeat">
+				<input ref="/data/num_repeat/repeats">
+					<label ref="jr:itext('question2/repeats-label')"/>
+				</input>
+			</repeat>
+		</group>
+	</h:body>
+</h:html>
```
Comment (benrudolph) [inline touchforms/backend/test_files/xforms/repeat.xml]: would be nice to make these files smaller/more concise so it's easier to tell what exactly they're testing

File: touchforms/backend/test_form_player_api.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    def setUp(self):
+
+        try:
+            f = open(os.path.join(CUR_DIR, 'test_files/xforms/question_types.xml'), 'r')
+            self.xform = f.read()
+            f = open(os.path.join(CUR_DIR, 'test_files/xforms/groups.xml'), 'r')
+            self.group_xform = f.read()
+            f = open(os.path.join(CUR_DIR, 'test_files/xforms/repeat.xml'), 'r')
+            self.repeat_xform = f.read()
+            f = open(os.path.join(CUR_DIR, 'test_files/xforms/nested.xml'), 'r')
```
Comment (benrudolph) [inline touchforms/backend/test_form_player_api.py]: i smell a utility function

File: touchforms/backend/test_form_player_api.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+            self.repeat_xform,
+            instance=None,
+            use_api=True,
+            **self.session_metadata
+        )
+
+        model = xform_session.fem
+        controller = xform_session.fec
+
+        assert(json_equals(xform_session.walk(), json.loads(WalkJson.walkToJson(model, controller))))
```
Comment (benrudolph) [inline touchforms/backend/test_form_player_api.py]: i believe self.assertEquals should ensure that dicts are equal (maybe in 2.5 we need to use self.assertDictEquals)

File: touchforms/backend/test_form_player_api.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            return False
+        if not json_equals(first.get(key), second.get(key)):
+            print "Comparing KEY: ", key
+            print "First: ", first.get(key)
+            print "===="
+            print "Second: ", second.get(key)
+            return False
+    return True
+
+def compare_list(first, second):
```
Comment (benrudolph) [inline touchforms/backend/test_form_player_api.py]: in general python has functions for this: https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertEqual

Comment (benrudolph) [inline touchforms/backend/test_form_player_api.py]: though you may need a utility for lists

File: touchforms/backend/xformplayer.py [hunk 5]
```diff
@@ -291,10 +296,15 @@ def output(self):
         return unicode(''.join(chr(b) for b in instance_bytes.tolist()), 'utf-8')
 
     def walk(self):
-        form_ix = FormIndex.createBeginningOfFormIndex()
-        tree = []
-        self._walk(form_ix, tree)
-        return tree
+        if self.use_api:
+            print "walking new"
```
Comment (benrudolph) [inline touchforms/backend/xformplayer.py]: nit: should be logging statements. probably `debug`

File: touchforms/backend/xformplayer.py [hunk 6]
```diff
@@ -291,10 +296,15 @@ def output(self):
...
+            return json.loads(WalkJson.walkToJson(self.fem, self.fec))
```
Comment (benrudolph) [inline touchforms/backend/xformplayer.py]: :joy:

File: touchforms/backend/xformplayer.py [hunk 7]
```diff
@@ -478,6 +489,10 @@ def back_event (self):
         return self._parse_current_event()
 
     def answer_question (self, answer, _ix=None):
+        if self.use_api:
```
Comment (benrudolph) [inline touchforms/backend/xformplayer.py]: do we need to do this for every call? i think a better pattern might be to make a base XFormSession class and then have a child class use the new api and another child class use the old way. that way there is less `if` branching in random parts of the code

**Comment by benrudolph** [comment]
the Pride strikes again! :leopard: :leopard: :leopard: 

i think we should hold off on merging this into the other branch until QA has finished and it's merged into master

**Comment by wpride** [comment]
@benrudolph I actually was not wanting this reviewed yet haha code is not nearly in a finished state but thanks for the pass.

**Comment by benrudolph** [comment]
ah gotcha sorry. just ping back when you're ready

**Comment by dimagimon** [comment]
There are 13 errors:

* touchforms/backend/test_form_player_api.py, line 4 - F401 'touchcare' imported but unused
* touchforms/backend/test_form_player_api.py, line 6 - F401 'AnswerQuestionJson' imported but unused
* touchforms/backend/test_form_player_api.py, line 6 - F401 'InstanceXmlUtil' imported but unused
* touchforms/backend/test_form_player_api.py, line 6 - F401 'PromptToJson' imported but unused
* touchforms/backend/test_form_player_api.py, line 7 - F401 'RuntimeException' imported but unused
* touchforms/backend/test_form_player_api.py, line 31 - E201 whitespace after '('
* touchforms/backend/test_form_player_api.py, line 112 - E303 too many blank lines (2)
* touchforms/backend/test_form_player_api.py, line 177 - E303 too many blank lines (2)
* touchforms/backend/test_form_player_api.py, line 192 - E713 test for membership should be 'not in'
* touchforms/backend/test_form_player_api.py, line 207 - E302 expected 2 blank lines, found 1
* touchforms/backend/test_form_player_api.py, line 208 - E703 statement ends with a semicolon
* touchforms/backend/xformplayer.py, line 599 - E302 expected 2 blank lines, found 1
* touchforms/backend/xformplayer.py, line 600 - E211 whitespace before '('

**Pull request closed without merging**